platform/o.n.agent/src/org/netbeans/agent/TransferHandlerActionsNonFinalTransformer.java

```diff
@@ -0,0 +1,107 @@
+ */
+public class TransferHandlerActionsNonFinalTransformer implements ClassFileTransformer {
+
+    private static final String DEBUG_DUMP_TRANSFORMED_CLASS = System.getProperty("org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.dumpPath");
+    private static final boolean DEBUG_DISABLE_TRANSFORMER = Boolean.getBoolean("org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.disable");
+
+    private static Logger getLog() {
+        return Logger.getLogger(TransferHandlerActionsNonFinalTransformer.class.getName());
+    }
+
+    @Override
+    public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+        if (DEBUG_DISABLE_TRANSFORMER) {
+            return null;
+        }
+        // We assume, that we are loaded early enough before AWT is initialzed,
+        // so we don't expect to need to retransform
+        if (classBeingRedefined != null) {
+            getLog().log(Level.WARNING, "Attempted retransform for: {0} (Ignored)", new Object[]{className});
+        }
+        // Only transform TransferHandler class - ignore all other classes
+        if ("javax/swing/TransferHandler".equals(className)) {
+            getLog().log(Level.INFO, "Transforming: {0}", new Object[]{className});
+            try {
+                ClassReader cr = new ClassReader(classfileBuffer);
+                ClassWriter cw = new ClassWriter(cr, ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
+                ClassVisitor cv = new ClassVisitor(Opcodes.ASM9, cw) {
+                    @Override
+                    public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+                        switch(name) {
+                            case "cutAction":
+                            case "copyAction":
+                            case "pasteAction":
+                                return super.visitField(access & (~ Opcodes.ACC_FINAL), name, descriptor, signature, value);
+                            default:
+                                return super.visitField(access, name, descriptor, signature, value);
+                        }
+                    }
+                };
+                cr.accept(cv, 0);
+                byte[] result = cw.toByteArray();
+                if (DEBUG_DUMP_TRANSFORMED_CLASS != null && !DEBUG_DUMP_TRANSFORMED_CLASS.isBlank()) {
+                    Files.write(Path.of(DEBUG_DUMP_TRANSFORMED_CLASS), result, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+                }
+                return result;
+            } catch (IOException | RuntimeException ex) {
+                getLog().log(Level.INFO, MessageFormat.format("Transforming: {0} failed", new Object[]{className}), ex);
+            } finally {
+                getLog().log(Level.INFO, "Transforming: {0} done", new Object[]{className});
```

Comment:
nitpick: would move this log line to before `return result;` so that it is not printed together with failed.

Reply:
JUL usage was removed here to not mess with its initialization by the IDE/unittesting infrastructure. Replacement call was moved as suggested.